src/client/desktop_gui/lib/shell_terminal.dart

```diff
@@ -0,0 +1,126 @@
+import 'package:flutter/services.dart';
+import 'package:flutter_pty/flutter_pty.dart';
+import 'package:flutter_riverpod/flutter_riverpod.dart';
+import 'package:xterm/xterm.dart';
+
+import 'settings.dart';
+
+const terminalFontSizeKey = 'terminalFontSize';
+
+class ShellTerminal extends ConsumerStatefulWidget {
+  final String instance;
+
+  const ShellTerminal({super.key, required this.instance});
+
+  @override
+  ConsumerState<ShellTerminal> createState() => _ShellTerminalState();
+}
+
+class _ShellTerminalState extends ConsumerState<ShellTerminal> {
+  final terminal = Terminal(maxLines: 10000);
+  final terminalController = TerminalController();
+
+  late double fontSize =
+      ref.read(settingsProvider).getDouble(terminalFontSizeKey) ?? 13;
+
+  void setFontSize(double size) {
+    size = size.clamp(5, 25);
+    setState(() => fontSize = size);
+    ref.read(settingsProvider).setDouble(terminalFontSizeKey, size);
+  }
+
+  void increaseFontSize() => setFontSize(fontSize + 0.5);
+
+  void decreaseFontSize() => setFontSize(fontSize - 0.5);
+
+  void resetFontSize() => setFontSize(13);
+
+  late final shortcuts = {
+    increaseFontSize.intent: [
+      const SingleActivator(LogicalKeyboardKey.numpadAdd, control: true),
+      const SingleActivator(LogicalKeyboardKey.equal, control: true),
+      const SingleActivator(LogicalKeyboardKey.add, control: true, shift: true),
+    ],
+    decreaseFontSize.intent: [
+      const SingleActivator(LogicalKeyboardKey.numpadSubtract, control: true),
+      const SingleActivator(LogicalKeyboardKey.minus, control: true),
+      const SingleActivator(LogicalKeyboardKey.minus,
+          control: true, shift: true),
+    ],
+    resetFontSize.intent: [
+      const SingleActivator(LogicalKeyboardKey.digit0, control: true),
+    ],
+  }.entries.fold(
+    // this goes from a map of {Intent1: [Shortcut1, Shortcut2], Intent2: [Shortcut3, Shortcut4]}
+    // to one of {Shortcut1: Intent1, Shortcut2: Intent1, Shortcut3: Intent2, Shortcut4: Intent2}
+    <SingleActivator, VoidCallbackIntent>{},
+    (map, entry) {
+      final intent = entry.key;
+      final activators = entry.value;
+      return {...map, for (final activator in activators) activator: intent};
+    },
+  );
+
+  @override
+  void initState() {
+    super.initState();
+    final pty = Pty.start(
+      'multipass',
+      arguments: ['shell', widget.instance],
+      environment: {
+        if (Platform.environment['SNAP_NAME'] == 'multipass')
+          "XDG_DATA_HOME": '${Platform.environment['SNAP_USER_DATA']!}/data',
+      },
+      columns: terminal.viewWidth,
+      rows: terminal.viewHeight,
+      ackRead: true,
+    );
+    pty.exitCode.then(
+      (code) {
+        if (code == 0) exit(0);
+        terminal.write('Process ended with exit code $code');
+      },
+    );
+    pty.output.listen((data) => terminal.write(utf8.decoder.convert(data)));
+    Stream.periodic(Duration.zero).listen((_) => pty.ackRead());
```

Comment:
I wish there was one...
First, to clarify why this is here: we could pass `ackRead: false` to `Pty.start`, and the the `output` stream will emit whenever there is data available from the process. But, in the case of a process that emits a lot of data (e.g. the `yes` command), the stream would emit so frequently that Flutter itself would not get a chance to run, and this would result in very slow rendering and basically no input processing, so doing a CTRL+C wouldn't stop the process. This is because Dart is single-threaded and  both the Flutter stuff and the output stream exist in the same event loop. And callbacks to stream subscriptions are executed on the micro-task queue, before other Flutter events.

Now, indeed, the ideal solution would be to have a mechanism to select between events emitted by the output stream and those of Flutter (preferably with a priority for the Flutter ones). But we don't even have anything to select on from the Flutter side i.e. there is no stream we can listen to that tells us when Flutter emits an event.

This solution simply repeatedly tells the pty that it can emit events from its output stream which, for some reason I don't fully understand, makes it so that the event loop also has a chance to process Flutter events. The duration doesn't have to be zero. I've tried with higher values, like every 100ms, and the terminal is still pretty responsive. `Duration.zero` simply feels the most responsive. But we can try higher values and maybe that will result in less CPU usage.